ec/ecresource/deploymentresource/deployment/v2/deployment_read_test.go

```diff
@@ -1723,3 +1728,100 @@ func Test_getDeploymentTemplateID(t *testing.T) {
 		})
 	}
 }
+
+func Test_ProcessSelfInObservability(t *testing.T) {
+	tests := []struct {
+		name                              string
+		deployment                        *Deployment
+		baseObservability                 *observabilityv1.Observability
+		expectedObservabilityDeploymentID *string
+		expectNonNilDiags                 bool
+	}{
+		{
+			name: "should noop if deployment is nil",
+		},
+		{
+			name:       "should noop if observability section is nil",
+			deployment: &Deployment{},
+		},
+		{
+			name: "should noop if observability deployment id is nil",
+			deployment: &Deployment{
+				Observability: &observabilityv1.Observability{},
+			},
+		},
+		{
+			name: "should not change the observability deployment id to self if it equals the deployment id and the configured value is not self",
+			deployment: &Deployment{
+				Id: "deployment-id",
+				Observability: &observabilityv1.Observability{
+					DeploymentId: ec.String("deployment-id"),
+				},
+			},
+			baseObservability: &observabilityv1.Observability{
+				DeploymentId: ec.String("deployment-id"),
+			},
+			expectedObservabilityDeploymentID: ec.String("deployment-id"),
+		},
+		{
+			name: "should set observability deployment id to self if it equals the deployment id and the configured value is self",
+			deployment: &Deployment{
+				Id: "deployment-id",
+				Observability: &observabilityv1.Observability{
+					DeploymentId: ec.String("deployment-id"),
+				},
+			},
+			baseObservability: &observabilityv1.Observability{
+				DeploymentId: ec.String("self"),
+			},
+			expectedObservabilityDeploymentID: ec.String("self"),
+		},
+		{
+			name: "should set observability deployment id to self if it equals the deployment id and the configured value is not set",
+			deployment: &Deployment{
+				Id: "deployment-id",
+				Observability: &observabilityv1.Observability{
+					DeploymentId: ec.String("deployment-id"),
+				},
+			},
+			baseObservability:                 &observabilityv1.Observability{},
+			expectedObservabilityDeploymentID: ec.String("self"),
+		},
+		{
+			name: "should not change the observability deployment id if it does not equal the deployment id",
+			deployment: &Deployment{
+				Id: "deployment-id",
+				Observability: &observabilityv1.Observability{
+					DeploymentId: ec.String("another-deployment-id"),
+				},
+			},
+			expectedObservabilityDeploymentID: ec.String("another-deployment-id"),
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			var obj types.Object
+			if tt.baseObservability != nil {
```

Comment:
Shall we cover the case when `base.Observability` is `Unknown`? (Sorry, if I missed something).

Reply:
Yep, added.